### PR TITLE
adjust check func that always print vulnerable problem

### PR DESCRIPTION
**Pull request by Ryuuuuu**
in this module, did not check version = 0. so that always say it is vulenerable.

i fixed that.

## Verification

i just add " || version.version == '0'" on code

### CODE REVIEW

**Comment by jvoisin** [comment]
Under what conditions would be version be `0`?

**Comment by bwatters-r7** [comment]
Oooof; if I'm reading this correctly, the `get_version` method sets the version as `nil` and returns false if there's a connection error, never resetting the `nil` value for version.  `Rex::Version.new(nil)` returns 0.
I think a better option would be to check the return value from the `get_version` method, but this looks like it would work.

```
  def get_version
    @version = nil

    begin
      res = send_request_cgi(
        'uri'           => normalize_uri(target_uri.path),
        'method'        => 'GET',
        'authorization' => @auth
      )
    rescue Rex::ConnectionError
      vprint_bad("#{peer} - Connection failed")
      return false
    end
```

```
msf6 exploit(linux/http/apache_couchdb_cmd_exec) > irb
[*] Starting IRB shell...
[*] You are in exploit/linux/http/apache_couchdb_cmd_exec

>> Rex::Version.new(nil)
=> Gem::Version.new("0")
```

You can verify this by running the module against a target without the port open:
```
[ruby-3.0.2](land-17401) tmoose@ubuntu:~/rapid7/metasploit-framework$ netstat -ant | grep 5984
[ruby-3.0.2](land-17401) tmoose@ubuntu:~/rapid7/metasploit-framework$ 
```

```
msf6 exploit(linux/http/apache_couchdb_cmd_exec) > show options

Module options (exploit/linux/http/apache_couchdb_cmd_exec):

   Name          Current Setting  Required  Description
   ----          ---------------  --------  -----------
   HttpPassword                   no        The password to login with
   HttpUsername                   no        The username to login as
   Proxies                        no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOSTS                         yes       The target host(s), see https://github.com/rapid7/metasploit-framework/wiki/Using-Metas
                                            ploit
   RPORT         5984             yes       The target port (TCP)
   SRVHOST       0.0.0.0          yes       The local host or network interface to listen on. This must be an address on the local
                                            machine or 0.0.0.0 to listen on all addresses.
   SRVPORT       8080             yes       The local port to listen on.
   SSL           false            no        Negotiate SSL/TLS for outgoing connections
   SSLCert                        no        Path to a custom SSL certificate (default is randomly generated)
   URIPATH                        no        The URI to use for this exploit to download and execute. (default is random)
   VHOST                          no        HTTP server virtual host


Payload options (linux/x64/shell_reverse_tcp):

   Name   Current Setting  Required  Description
   ----   ---------------  --------  -----------
   LHOST                   yes       The listen address (an interface may be specified)
   LPORT  4444             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   Automatic



View the full module info with the info, or info -d command.

msf6 exploit(linux/http/apache_couchdb_cmd_exec) > set rhost 127.0.0.1
rhost => 127.0.0.1
msf6 exploit(linux/http/apache_couchdb_cmd_exec) > set payload linux/x64/meterpreter/reverse_tcp 
payload => linux/x64/meterpreter/reverse_tcp
msf6 exploit(linux/http/apache_couchdb_cmd_exec) > set verbose true
verbose => true                                   
msf6 exploit(linux/http/apache_couchdb_cmd_exec) > check
                                                  
[-] The connection was refused by the remote host (127.0.0.1:5984).
[-] 127.0.0.1:5984 - No response, check if it is CouchDB. 
[*] Found CouchDB version 0                       
[*] 127.0.0.1:5984 - The target appears to be vulnerable.
msf6 exploit(linux/http/apache_couchdb_cmd_exec) > 
```

In this case, version `0` passes this check:
```
    return CheckCode::Appears if version < Rex::Version.new('1.7.0') || version.between?(Rex::Version.new('2.0.0'), Rex::Version.new('2.1.0'))

```

**Comment by bwatters-r7** [comment]
On second thought, this works better than checking the return value..... 
```
      @version = res_json['version'] if res_json['version']
      return true
```
That conditional might not set the version, and still return true.

**Comment by bwatters-r7** [comment]
Forgive the steam-of-consciousness spamming, but I think the better option now is to just define the default version as -1.  In that case, `Rex::Version` fails and the module crashes if it tries to convert it to a version.  We can then add a check for -1 to avoid the crash and show that the cause was a version error.
TL;DR: always choose default values that are invalid.

Changing the default value to be -1 (an invalid value):
```
msf6 exploit(linux/http/apache_couchdb_cmd_exec) > check
                               
[-] The connection was refused by the remote host (127.0.0.1:5984).
[-] 127.0.0.1:5984 - No response, check if it is CouchDB. 
[-] Exploit failed: ArgumentError Malformed version number string -1
[-] 127.0.0.1:5984 - Check failed: The state could not be determined.

```
Adding a check for -1 as the version value:
```
msf6 exploit(linux/http/apache_couchdb_cmd_exec) > check
                               
[-] The connection was refused by the remote host (127.0.0.1:5984).
[-] 127.0.0.1:5984 - No response, check if it is CouchDB. 
[-] Exploit failed: ArgumentError Malformed version number string -1
[-] 127.0.0.1:5984 - Check failed: The state could not be determined.
msf6 exploit(linux/http/apache_couchdb_cmd_exec) > 

```

File: modules/exploits/linux/http/apache_couchdb_cmd_exec.rb [hunk 1]
```diff
@@ -70,7 +70,7 @@ def post_auth?
   def check
     get_version
     version = Rex::Version.new(@version)
-    return CheckCode::Unknown if version.version.empty?
+    return CheckCode::Unknown if version.version.empty? || version.version == '0'
```
Comment (bwatters-r7) [inline modules/exploits/linux/http/apache_couchdb_cmd_exec.rb:73]:
```suggestion
    return CheckCode::Unknown if version.version.empty?
```

File: modules/exploits/linux/http/apache_couchdb_cmd_exec.rb [hunk 2]
```diff
@@ -70,7 +70,7 @@ def post_auth?
   def check
     get_version
     version = Rex::Version.new(@version)
```
Comment (bwatters-r7) [inline modules/exploits/linux/http/apache_couchdb_cmd_exec.rb:72]:
```suggestion
    return CheckCode::Unknown if @version < 0
    version = Rex::Version.new(@version)
```

**Review by bwatters-r7 — COMMENTED** [review]
I could not make a suggested change to the version default on line 134, but rather than setting to `nil` is should be set to -1.

I can make these changes and put up a PR, but you found it and properly understood the issue, @Ryuuuuu, so if you want to make the changes and get credit for it, please do so.

**Pull request closed without merging**